### ISSUE TITLE
[BUG] - problem installing packages from git into environment

### ISSUE THREAD

**Issue by rsignell-usgs** [Quansight/qhub#1305]
### OS system and architecture in which you are running QHub

qhub 0.4.2

### Expected behavior

environments containing lines like this work:
```yaml
- hvplot
- pip
- pip:
  - git+https://github.com/zarr-developers/numcodecs
  - git+https://github.com/observingClouds/xbitinfo
```

### Actual behavior

![2022-06-02_15-47-55](https://user-images.githubusercontent.com/1872600/171726309-eba7c2c4-85c3-45b8-92e6-3c8060d393a5.png)

### How to Reproduce the problem?

```yaml
name: bitinfo
channels:
- conda-forge
dependencies:
- python=3.10
- requests
- aiohttp
- intake-xarray
- julia
- pyjulia
- matplotlib-base
- numpy
- pooch
- xarray
- dask
- distributed
- ipykernel
- ipywidgets
- s3fs
- netcdf4
- zarr
- cfgrib
- prefect[version='>=1.0.0']
- graphviz
- pytest
- nb_conda_kernels
- cmcrameri
- tqdm
- asv
- sphinx!=4.4.0
- sphinxcontrib-napoleon
- sphinx-copybutton
- sphinx-book-theme
- myst-nb
- ca-certificates
- certifi
- openssl
- datashader
- geoviews
- hvplot
- pip
- pip:
  - git+https://github.com/zarr-developers/numcodecs
  - git+https://github.com/observingClouds/xbitinfo
 ```

### Compute environment

AWS

**Comment by viniciusdc** [comment]
Hi @rsignell-usgs thanks for opening this issue, regarding the problem itself this is more related to[ conda-store](https://github.com/Quansight/conda-store) than qhub, so if you don't mind I will raise an issue to address this in there as well.

The env build is falling because conda-store has an [inbuilt validator](https://github.com/Quansight/conda-store/blob/11f8538a0f86c24f55e02bdc41de0fdf18639ce6/conda-store-server/conda_store_server/schema.py#L142) to inspect the pip/conda dependencies. And based on the [Requirement.parse()](https://setuptools.pypa.io/en/latest/pkg_resources.html#requirements-parsing) specifications, the `"git+https://github.com/zarr-developers/numcodecs"` string does not satisfies the PEP508 specifiers. 

So, in short, conda-store needs to fix the parsing for such cases, we will try to address this asap. Thanks again for rasing this :slightly_smiling_face:

**Comment by viniciusdc** [comment]
Hi @peytondmurray, what is the current status on this?

**Comment by peytondmurray** [comment]
@viniciusdc See [this comment](https://github.com/Quansight/conda-store/issues/317) for a bit more info, but in short packages installed via pip from VCS must follow the [PEP440 format](https://peps.python.org/pep-0440/#direct-references), e.g.

```
# environment.yml
name: test-env
dependencies:
- hvplot
- pip
- pip:
  - numcodecs @ git+https://github.com/zarr-developers/numcodecs@main
  - xbitinfo @ git+https://github.com/observingClouds/xbitinfo@main
```

where the pip dependencies are specified like

```
<name of package to use> @ git+https://<github repo URL>@<commit hash or tag>
```

In the example above this _should_ work, except the current development version of `xbitinfo` depends on a specific version of `numcodecs`. That's is an issue with those specific pip dependencies, though, not qhub or conda-store.

**Comment by viniciusdc** [comment]
Yup, @peytondmurray thats the expected format we do expect in the validation. Have you tested if the above environment builds using the expected pip syntax? This would be enough to close the issue and open a doc update one.

**Comment by peytondmurray** [comment]
Well, the development version of `xbitinfo` depends on the [development version of numcodecs](https://github.com/observingClouds/xbitinfo/blob/main/requirements.txt). Specifying both dependencies explicitly breaks pip's dependency solver:

```
The conflict is caused by:
    The user requested numcodecs 0.10.0a4.dev5 (from git+https://github.com/zarr-developers/numcodecs@main)
    xbitinfo 0.0.1.post13 depends on numcodecs 0.10.0a4.dev5 (from git+https://github.com/zarr-developers/numcodecs)

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict


failed

CondaEnvException: Pip failed
```

Specifying `xbitinfo` alone also doesn't work, because again, pip's dependency solver seems unable to handle development versions of packages:

```yaml
# environment.yaml
channels:
- conda-forge
dependencies:
- gcc
- cython
- hvplot
- pip
- pip:
  - numcodecs @ git+https://github.com/zarr-developers/numcodecs@main
  - xbitinfo @ git+https://github.com/observingClouds/xbitinfo@main
- ipykernel
name: test-environment
prefix: null
```

:point_up: results in :point_down: 

```
Looking for: ['gcc', 'cython', 'hvplot', 'pip', 'ipykernel']


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
Installing pip dependencies: ...working... Pip subprocess error:
  Running command git clone --filter=blob:none --quiet https://github.com/zarr-developers/numcodecs /tmp/pip-install-wemh84w2/numcodecs_04b010a4fbf0486abef69fe5d3cb72d7
  Running command git submodule update --init --recursive -q
  Running command git clone --filter=blob:none --quiet https://github.com/observingClouds/xbitinfo /tmp/pip-install-wemh84w2/xbitinfo_9515a463c4bd4ea58e063e44be88398b
  Running command git clone --filter=blob:none --quiet https://github.com/zarr-developers/numcodecs /tmp/pip-install-wemh84w2/numcodecs_fa43f2582edc4eafaac6d501bb4caac1
  Running command git submodule update --init --recursive -q
ERROR: Cannot install -r /tmp/tmpqf8b31p8/condaenv.0r805g0r.requirements.txt (line 2) and numcodecs 0.10.0a4.dev5 (from git+https://github.com/zarr-developers/numcodecs@main) because these package versions have conflicting dependencies.
ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```
---
Inside `conda-store`, I tried installing a different package from source:

```yaml
# environment.yaml
dependencies:
- pip
- pip:
  - tqdm @ git+https://github.com/tqdm/tqdm@master
- ipykernel
name: test-environment-conda-store
```

But ran into another seemingly unrelated issue: https://github.com/Quansight/conda-store/issues/299. However, installing the environment above from `conda` running on my desktop succeeded, so I think this is just a conda-store problem right now.
